src/testing/slim_check/sampleable.lean

```diff
@@ -75,10 +84,98 @@ class sampleable :=
 
 attribute [instance, priority 100] has_well_founded_of_has_sizeof default_has_sizeof
 attribute [instance, priority 200] sampleable.wf
+
+/-- `sampleable_functor F` makes it possible to create samples of and
+shrink `F α` given a sampling function and a shrinking function for
+arbitrary `α` -/
+class sampleable_functor (F : Type u → Type v) [functor F] :=
+[wf : Π α [has_sizeof α], has_sizeof (F α)]
+(sample [] : ∀ {α}, gen α → gen (F α))
+(shrink : ∀ α [has_sizeof α], shrink_fn α → shrink_fn (F α))
+(p_repr : ∀ α, has_repr α → has_repr (F α))
+
+/-- `sampleable_bifunctor F` makes it possible to create samples of
+and shrink `F α β` given a sampling function and a shrinking function
+for arbitrary `α` and `β` -/
+class sampleable_bifunctor (F : Type u → Type v → Type w) [bifunctor F] :=
+[wf : Π α β [has_sizeof α] [has_sizeof β], has_sizeof (F α β)]
+(sample [] : ∀ {α β}, gen α → gen β → gen (F α β))
+(shrink : ∀ α β [has_sizeof α] [has_sizeof β], shrink_fn α → shrink_fn β → shrink_fn (F α β))
+(p_repr : ∀ α β, has_repr α → has_repr β → has_repr (F α β))
+
 export sampleable (sample shrink)
 
+/-- This function helps infer the proxy representation and
+interpretation in `sampleable_ext` instances. -/
+meta def sampleable.mk_trivial_interp : tactic unit :=
+tactic.refine ``(id)
+
+/-- `sampleable_ext` generalizes the behavior of `sampleable`
+and makes it possible to express instances for types that
+do not lend themselves to introspection, such as `ℕ → ℕ`.
+If we test a quantification over functions the
+counter-examples cannot be shrunken or printed meaningfully.
+
+For that purpose, `sampleable_ext` provides a proxy representation
+`proxy_repr` that can be printed and shrunken as well
+as interpretted (using `interp`) as an object of the right type. -/
+class sampleable_ext (α : Sort u) :=
+(proxy_repr : Type v)
+[wf : has_sizeof proxy_repr]
+(interp [] : proxy_repr → α . sampleable.mk_trivial_interp)
+[p_repr : has_repr proxy_repr]
+(sample [] : gen proxy_repr)
+(shrink : shrink_fn proxy_repr)
+
+attribute [instance, priority 100] sampleable_ext.p_repr sampleable_ext.wf
+
 open nat lazy_list
 
+section prio
+
+open sampleable_ext
+
+set_option default_priority 50
+
+instance {α} [sampleable α] [has_repr α] : sampleable_ext α :=
+{ proxy_repr := α,
+  sample := sampleable.sample α,
+  shrink := shrink }
+
+instance sampleable.functor {α} {F} [functor F] [sampleable_functor F] [sampleable α] : sampleable (F α) :=
+{ wf := _,
+  sample := sampleable_functor.sample F (sampleable.sample α),
+  shrink := sampleable_functor.shrink α sampleable.shrink }
+
+instance sampleable.bifunctor {α β} {F} [bifunctor F] [sampleable_bifunctor F] [sampleable α] [sampleable β] : sampleable (F α β) :=
+{ wf := _,
+  sample := sampleable_bifunctor.sample F (sampleable.sample α) (sampleable.sample β),
+  shrink := sampleable_bifunctor.shrink α β sampleable.shrink sampleable.shrink }
+
+set_option default_priority 100
```

Comment:
I think we no longer need to do this!

Reply:
Isn't the new feature about the instances created when declaring new classes? The generic instances that this setting is intended for are a different situation.